### ISSUE TITLE
support plink as SSH/SFTP client on Windows

### ISSUE THREAD

**Issue by burritts** [restic/restic#818]
I've been working with restic 0.4.0, compiled with go1.7.5 on windows/amd64.  My initial attempts to access an existing remote repo using sftp failed (predictably) with an error like:
unable to open repo at sftp://user@host//path/to/repo: cmd.Start: exec: "ssh": executable file not found in %PATH%

I experimented with using plink.exe and was able to to get things working by hacking the sftp backend to construct the ssh command line like so:
path/to/plink.exe -s -l {username} -P {port} -i {path/to/myprivatekey.ppk} {host} sftp
or when using pagent:
path/to/plink.exe -s -l {username} -P {port} -agent {host} sftp

More info on plink [here](https://the.earth.li/~sgtatham/putty/0.68/htmldoc/Chapter7.html#plink)

plink is a single-file exe download, and is more acceptable to most Windows users (I think) than installing cygwin, or using the beta windows 10 Linux subsystem.

I'm willing to work on a contribution to support this, if the community is interested.  If so, I'd appreciate some input on how best to configure/expose plink vs standard ssh behaviour.  My current prototype uses query string terms on the sftp repository URL, like this:
`restic -r "sftp://user@host//tmp/backup?plink=/path/plink.exe&ppk=/path/to/privatekey.ppk" ...`
or 
`restic -r "sftp://user@host//tmp/backup?plink=/path/plink.exe&agent" ...`

If plink can be found on the path, the url could be as simple as:
`restic -r "sftp://user@host//tmp/backup?plink&ppk=/path/to/privatekey.ppk" ...`

**Comment by xor-gate** [comment]
In my opinion, good Golang applications should avoid `os.Exec` whenever possible (even through libraries which should be reviewed for `os.Exec` calls). Users can get into unwanted surprises when bad libraries just call `os.Exec` all the way.

**Comment by dirk777** [comment]
I didn't get your idea to work so far but your approach using plink sounds good. In particular it could be a solution for my problem using restic and openssh under windows as reported on issue #780 . So if you provide more information to get your solution run, i can test it. I highly appreciate a working and stable environment under windows.

**Comment by burritts** [comment]
I will work towards creating a PR for the changes, but it will be a couple of weeks before I'll be able to get to it.  I had hoped that the project committers (is that @fd0 ?) would let me know if they are OK with using query string style terms to customize the sftp URL for plink.  @dirk777 are you willing to build restic and test the changes once I create a fork and make them available?

**Comment by fd0** [comment]
Hey, thanks for raising this issue. I don't have much time at the moment, so I didn't have an in-depth look at your proposal yet. Please do not work on it before we agree on a config format.

I think the URL style configuration for the backend is complicated as is, and I must admit that I don't like making it even more complicated by adding URL parameters. Let me think about this for a bit, I'll come back to you. Thanks!

**Comment by burritts** [comment]
I wasn't thrilled with the URL parameters, either.  It seemed the least disruptive thing I could do that was halfway consistent with the current URL-ish scheme. I'm happy to hold off until the best approach is clear.  Perhaps it is time to consider some scheme where configurations for backends are saved in individual configuration files, or named profiles or some such.

**Comment by fd0** [comment]
I'll see what I can do :)

**Comment by fd0** [comment]
So, I'd like to propose adding free-form options (like `ssh -o`) to configure this detail for the sftp backend. For this particular use case, users can then call restic as follows to:
```
$ restic -o sftp.command="path/to/plink.exe -s -l user -P 22 -i path/to/myprivatekey.ppk host sftp" --repo sftp:host:/path/to/repo
```
Then restic won't construct the command itself, but rather use the provided command. What do you think?

**Comment by xor-gate** [comment]
I totaly agree @fd0 this is also how rsync can be set to a different ssh executable.

**Comment by burritts** [comment]
That looks fine to me. Someone will no doubt have trouble figuring out how to escape spaces in folder names, so consider documenting how to escape spaces, or  recommend "don't do that".

**Comment by fd0** [comment]
Spaces for which folder name? I guess that the executable doesn't need a path in most cases (since it will probably be in `$PATH`), and for the repo path you can just enclose it in single or double quotes. When the command line for starting the sftp client is too complicated, users can even write a short script (or batch file) and reference that in the restic command line...

**Comment by burritts** [comment]
Sorry, that comment was written in haste and wasn't very clear. To elaborate: either the path to the executable or to the private key ppk file could have folders with spaces in them.  The executable may usually be in the path, but if we need to support cases where it is not, then folders with spaces may need to be supported. The ppk file is not going to be found on the path.  Consider this: -o sftp.command="path with spaces/to/plink.exe -s -l user -P 22 -i path with spaces/to/myprivatekey.ppk host sftp".  Quotes have already been used to form the value on the right side of the equals sign.  How should the spaces within the path terms inside the quoted value be escaped?  Maybe URL encode the value? 

I get your point about the user creating a script or batch file to start the sftp client, and that appears to avoid the issue nicely.  We'll have to verify that windows hooks up I/O correctly when plink.exe (or whatever) is invoked from a bat file.  It may work fine, I have no idea.  I can try it, but it will be next week before I can get to it.

I hope I am not being too annoying by seeming to obsess over folders with spaces in them. Personally, I think it is just a bad idea, but Windows users seem determined to do it, and it often seems to get forgotten and cause problems.  I just wanted to make sure it was considered, and either supported somehow, or explicitly not supported.

**Comment by burritts** [comment]
Using a bat file to run plink appears to work well.  To test this I made a purely experimental change in sftp.go to create the ssh command like so:

```
	if prog := os.Getenv("RESTIC_SFTP"); prog != "" {
		fmt.Printf("using sftp cmd %+v\n", prog)
		return exec.Command(prog)
	}
```

Note: I am **not** suggesting use of an environment variable to configure the sftp backend program in real life, this was just a simple way to experiment with the code.

I set the environment variable to the path to a bat file "/tools/runplink.bat" with contents as below:

```
@echo off
/tools/plink.exe -s -l bob -i /tools/bobskey.ppk hostname sftp
```

`@echo off` is important - without it commands executed are echoed to stdout, and restic hangs, presumably because it receives unexpected input.

restic appears to work fine.  From a Windows 10 VM I init'ed a repo on a server running Ubuntu 14.04, then created a backup of about 2.1 GB.  I did not see the hangs reported in #780 

My experiment is equivalent to `-o "sftp.command=/tools/runplink.bat"`. 

I think it would be reasonable for restic to support sftp.command of the form "program [args...]" where args are optional, and if args are provided they are white space delimited and are used as-is.  No escaping, no unnecessary complications in the restic source.  If that isn't good enough (for example because of spaces in folder names), the user can write a script or bat file, exactly as fd0 suggests.
